registry.go

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+// A Registry is a scoped metric registry. All metrics created with a Registry
+// will have the Registry's labels appended to them.
+type Registry struct {
+	core        *coreRegistry
+	constLabels Labels
+}
+
+func newRegistry(core *coreRegistry, labels Labels) *Registry {
+	return &Registry{
+		core:        core,
+		constLabels: labels,
+	}
+}
+
+// Labeled creates a new Registry with new constant labels appended to the
+// current Registry's existing labels. Label names and values are
+// automatically scrubbed, with invalid characters replaced by underscores.
+func (r *Registry) Labeled(ls Labels) *Registry {
+	if r == nil {
+		return nil
+	}
+	newLabels := make(Labels, len(r.constLabels)+len(ls))
+	for k, v := range r.constLabels {
```

Comment:
If this has conflicts with the existing const labels we won't notice until we attempt to create a metric right?

Reply:
Correct. This seems useful to me - even today, we frequently intentionally overwrite metrics (imagine the "component" tag, for example).

Reply:
Oh, we're overwriting, riiiiight, sgtm.